tutorial.md

```diff
@@ -0,0 +1,378 @@
+
+        for byte in buf {
+            while self.FR.matches_all(FR::TXFF::SET) {
+                bsp::nop();
+            }
+
+            self.DR.write(DR::DATA.val(*byte as u32));
+        }
+
+        Ok(buf.len())
+    }
+}
+
+impl io::Read for PL011Uart {
+    type Err = KernelError;
+    
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize, Self::Err> {
+        for byte in buf { 
+            while self.FR.matches_all(FR::RXFE::SET) {
+                bsp::nop();
+            }
+    
+            *byte = self.DR.read(DR::DATA) as u8;
+        }
+
+        Ok(buf.len())
+    }
+}
+```
+
+FR - flag register helps us to detect if device is busy at RW operations, so we wait for device is available for them. There is still work to do. We need to check hardware errors, but that is a topic for a separate article. <br/>
+<br/>
+Both drivers are ready, now we need to add opportunity for kernel to initialize them:
+
+```Rust
+static mut UART: uart::PL011Uart = uart::PL011Uart{};
+static mut GPIO: gpio::GPIO = gpio::GPIO{};
+
+pub fn console() -> &'static mut impl io::Console {
+    unsafe { &mut UART }
+} 
+
+pub fn drivers() -> [&'static dyn Driver; 2] {
+    unsafe { [&GPIO, &UART] }
+}
+
+pub fn post_init() {
+    unsafe { GPIO.map_pl011_uart() }
+}
+```
+I used `unsafe` here, because compiler does not allow to use mutable statics for concurency safety reasons. Since we use only one core, we can't have such problems. Of cource it needs some refactoring in future.
+
+We done writing board-specific code. Now we need to use it in kernel:
+
+```Rust
+pub fn init() -> ! {
+    for driver in bsp::drivers().iter() {
+        if let Err(_) = driver.init() {
+            panic!("failed to load driver: {:?}", driver.name())
+        }
+    }
+    bsp::post_init();
+    
+    kernel_main()
+}
+
+fn kernel_main() -> ! {
+    let mut data = [0u8];
+
+    // wait until user hits Enter
+    loop {
+        bsp::console().read(&mut data);
+        if data[0] as char == '\n' {
+            break;
+        }
+    }
+
+    // echo the input
+    loop {
+        bsp::console().read(&mut data);
+        bsp::console().write(&data);
+    }
+}
+```
+
+This little program waits for user until he hits Enter key. After this kernel will echo the user's input.<br/>
+Full source code can be found here. You can find there instruction how to set your environment correctly and how to build the kernel. It is very simple to do it using Rust tools.
+
+## Conclusions
+Rust is a system language such as C++ or C. Large crates registry allowed us to write this little kernel without any raw assembler instructions. To achieve this we used `cortex-a` crate that adds zero-cost abstractions over the registers and assembler commands. Another crate called `register` helped us to avoid all the work with memory mappings. All we needed is just to specify the base address and offsets.
+All this facts make Rust suitable system level language.
```

Comment:
I liked the article in general, but I miss some pros or reasons why I should switch to Rust. Moreover, it would be great if you will make the Conclusions section in the form of a statements list.